libp2p/stream/bufferstream.nim

```diff
@@ -0,0 +1,277 @@
+## Nim-LibP2P
+## Copyright (c) 2018 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+## This module implements an asynchronous buffer stream 
```

Comment:
I see a lot of value if we settle on a single stream abstraction that can be used across all of our libraries. At the moment, I'm pushing for [nim-faststreams](https://github.com/status-im/nim-faststreams) to become such a package, but admittedly it still needs to be developed further for the asynchronous use cases and I need to document it much better.

I believe I have figured out how to handle in a very efficient way many of the typical problems when dealing with streams (e.g. layering streams on top of each other, ensuring correct buffer management for block ciphers and compressors, dealing with back pressure, handling delayed writes such as length prefixes, etc). Let's talk in Gitter to discuss what is needed in LibP2P.

The benefit of having a single stream abstraction is that we'll be able to code our entire networking stack in a way that most of the data will be consumed and written directly from the streams without any intermediate copies and large memory allocations. Since our serialization and compression libraries (such as SSZ, JSON and eventually ProtoBuf) are based on the faststreams interface, it would be possible to stream large responses from RPC calls while having a very low memory footprint on our target devices.

Reply:
again, after the prototype is done - divide and conquer.

Reply:
My message is an invitation for a conversation, not a call for a rewrite of the prototype :)

Reply:
Super open to standardize on a streaming interface. The current LPStream was inspired by Chronos streams, but there might be a better more general API out there.